akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameOutHandler.scala

```diff
@@ -105,16 +104,18 @@ private[http] class FrameOutHandler(serverSide: Boolean, _closeTimeout: FiniteDu
     /**
      * we have sent out close frame and wait for peer to sent its close frame
      */
-    private class WaitingForPeerCloseFrame(timeout: Timestamp = closeTimeout) extends InHandler with ProcotolExceptionHandling {
+    private class WaitingForPeerCloseFrame(deadline: Deadline = closeDeadline()) extends InHandler with ProcotolExceptionHandling {
       override def onPush() =
         grab(in) match {
           case Tick =>
-            if (timeout.isPast) completeStage()
-            else pull(in)
+            if (deadline.isOverdue()) {
+              if (log.isDebugEnabled) log.debug(s"Peer did not acknowledge CLOSE frame after ${_closeTimeout}, closing underlying connection now.")
+              completeStage()
+            } else pull(in)
           case PeerClosed(code, reason) =>
             if (serverSide) completeStage()
             else {
-              setHandler(in, new WaitingForTransportClose())
+              setHandler(in, new WaitingForTransportClose(deadline))
```

Comment:
Hmm, so previously we had `_closeTimeout` for the peer to send its close frame, and another `_closeTimeout` for the transport to close, while now we expect to see both within one `_closeTimeout`.

Is that what we want?

Reply:
I think we can redefine it's meaning as we see fit. Previously, the actual delay was pretty confusing because it didn't seem to correspond to any of the configured values because it was twice as long as expected.

Reply:
I like having separate timeouts for each phase: if each phase 'normally' takes 30ms and 'exceptionally' 2s, you can set the timeout to 2s. Then, when the first action completes normally, the second one will be aborted within 30ms+2s. When using just one timeout, you'd have to set it to 4s, and even when the first half completes normally within 30ms you won't abort the second half before the full 4s have expired.

But it's not important :D

Reply:
> But it's not important :D

Yeah, the question would be why you would be interested in each of those phases. In the end, what you care about is that a connection that might be hung actually is torn down after the deadline.